File: app/src/main/java/com/druger/aboutwork/presenters/CompanyDetailPresenter.java
package com.druger.aboutwork.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.druger.aboutwork.db.FirebaseHelper;
import com.druger.aboutwork.interfaces.view.CompanyDetailView;
import com.druger.aboutwork.model.CompanyDetail;
import com.druger.aboutwork.model.Review;
import com.druger.aboutwork.model.User;
import com.druger.aboutwork.rest.RestApi;
import com.druger.aboutwork.utils.rx.RxUtils;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;
import timber.log.Timber;

/**
 * Created by druger on 01.05.2017.
 */

@InjectViewState
public class CompanyDetailPresenter extends BasePresenter<CompanyDetailView>
        implements ValueEventListener {

    @Inject
    public CompanyDetailPresenter(RestApi restApi) {
        this.restApi = restApi;
    }

    private FirebaseAuth auth;
    private FirebaseAuth.AuthStateListener authListener;

    private DatabaseReference dbReference;
    private ValueEventListener valueEventListener;

    private List<Review> reviews = new ArrayList<>();

    public void getReviews(String companyID, int currentPage) {
        getViewState().showProgressReview();
        dbReference = FirebaseDatabase.getInstance().getReference();
        Query reviewsQuery = FirebaseHelper.INSTANCE.getReviewsForCompany(dbReference, companyID, currentPage);
        reviewsQuery.addValueEventListener(this);
    }

    @Override
    public void onDataChange(DataSnapshot dataSnapshot) {
        fetchReviews(dataSnapshot);
    }

    private void fetchReviews(DataSnapshot dataSnapshot) {
        reviews.clear();

        for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
            final Review review = snapshot.getValue(Review.class);
            Query queryUser = FirebaseHelper.INSTANCE.getUser(dbReference, review.getUserId());
            valueEventListener = new ValueEventListener() {
                @Override
                public void onDataChange(DataSnapshot dataSnapshot) {
                    if (dataSnapshot.exists()) {
                        for (DataSnapshot data : dataSnapshot.getChildren()) {
                            User user = data.getValue(User.class);
                            review.setName(user.getName());
                            getViewState().updateAdapter();
                        }
                    }
                }

                @Override
                public void onCancelled(DatabaseError databaseError) {

                }
            };
            queryUser.addValueEventListener(valueEventListener);
            review.setFirebaseKey(snapshot.getKey());
            reviews.add(review);
        }
        getViewState().hideProgressReview();
        getViewState().showReviews(reviews);
    }

    @Override
    public void onCancelled(DatabaseError databaseError) {
        Timber.e(databaseError.getMessage());
    }

    public void removeListeners() {
        if (dbReference != null) {
            dbReference.removeEventListener(this);
            if (valueEventListener != null) {
                dbReference.removeEventListener(valueEventListener);
            }
        }
    }

    public void getCompanyDetail(String companyID) {
        getViewState().showErrorScreen(false);
        getViewState().showProgress(true);
        requestCompanyDetail(companyID);
    }

    private void requestCompanyDetail(String companyID) {
        Disposable request = restApi.getCompany().getCompanyDetail(companyID)
                .compose(RxUtils.httpSchedulers())
                .subscribe(this::successGetCompanyDetails, this::handleError);

        unSubscribeOnDestroy(request);
    }

    private void successGetCompanyDetails(CompanyDetail companyDetail) {
        getViewState().showProgress(false);
        getViewState().showCompanyDetail(companyDetail);
    }

    @Override
    protected void handleError(Throwable throwable) {
        super.handleError(throwable);
        getViewState().showProgress(false);
        getViewState().showErrorScreen(true);
    }

    public void checkAuthUser() {
        auth = FirebaseAuth.getInstance();
        initAuthListener();
        auth.addAuthStateListener(authListener);
    }

    private void initAuthListener() {
        authListener = firebaseAuth -> {
            FirebaseUser user = firebaseAuth.getCurrentUser();
            if (user == null)
                getViewState().showAuth();
            else getViewState().addReview();
        };
    }

    public void removeAuthListener() {
        if (auth != null) {
            auth.removeAuthStateListener(authListener);
        }
    }
}

File: app/src/main/java/com/druger/aboutwork/presenters/CompanyDetailPresenter.kt
package com.druger.aboutwork.presenters

import com.druger.aboutwork.R
import com.druger.aboutwork.db.FirebaseHelper
import com.druger.aboutwork.enums.FilterType
import com.druger.aboutwork.interfaces.view.CompanyDetailView
import com.druger.aboutwork.model.CompanyDetail
import com.druger.aboutwork.model.Review
import com.druger.aboutwork.model.User
import com.druger.aboutwork.rest.RestApi
import com.druger.aboutwork.utils.rx.RxUtils
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import moxy.InjectViewState
import timber.log.Timber
import java.util.*
import javax.inject.Inject

/**
 * Created by druger on 01.05.2017.
 */

@InjectViewState
class CompanyDetailPresenter @Inject
constructor(restApi: RestApi) : BasePresenter<CompanyDetailView>(), ValueEventListener {

    private lateinit var auth: FirebaseAuth
    private var authListener: FirebaseAuth.AuthStateListener? = null

    private var dbReference: DatabaseReference? = null
    private var valueEventListener: ValueEventListener? = null

    private val reviews = ArrayList<Review>()
    private var position: String = ""
    private var city: String = ""
    private var filterType = FilterType.RATING

    init {
        this.restApi = restApi
    }

    fun getReviews(companyID: String) {
        viewState.showProgress(true)
        dbReference = FirebaseDatabase.getInstance().reference
        dbReference?.let {
            val reviewsQuery = FirebaseHelper.getReviewsForCompany(it, companyID)
            reviewsQuery.addValueEventListener(this)
        }
    }

    override fun onDataChange(dataSnapshot: DataSnapshot) {
        fetchReviews(dataSnapshot)
    }

    private fun fetchReviews(dataSnapshot: DataSnapshot) {
        reviews.clear()

        if (dataSnapshot.exists()) {
            for (snapshot in dataSnapshot.children) {
                val review = snapshot.getValue(Review::class.java)
                dbReference?.let { db ->
                    val queryUser = review?.userId?.let { FirebaseHelper.getUser(db, it) }
                    valueEventListener = object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            if (dataSnapshot.exists()) {
                                for (data in dataSnapshot.children) {
                                    val user = data.getValue(User::class.java)
                                    review?.name = user?.name
                                    viewState.updateAdapter()
                                }
                            }
                        }

                        override fun onCancelled(databaseError: DatabaseError) {
                            Timber.e(databaseError.message)
                            viewState.showProgress(false)
                        }
                    }
                    valueEventListener?.let { queryUser?.addValueEventListener(it) }
                }
                review?.firebaseKey = snapshot.key
                review?.let { reviews.add(it) }
            }
            viewState.showProgress(false)
            reviews.reverse()
            viewState.showReviews(reviews)
            checkFilterSettings()
        } else {
            viewState.showProgress(false)
            viewState.showEmptyReviews()
        }
    }

    private fun checkFilterSettings() {
        if (position.isNotEmpty() || city.isNotEmpty()) {
            filterReviews(filterType, position, city)
            viewState.setFilterIcon(R.drawable.ic_filter_applied)
        } else {
            viewState.setFilterIcon(R.drawable.ic_filter)
        }
    }

    override fun onCancelled(databaseError: DatabaseError) {
        Timber.e(databaseError.message)
        viewState.showProgress(false)
    }

    fun removeListeners() {
        dbReference?.removeEventListener(this)
        valueEventListener?.let { dbReference?.removeEventListener(it) }
    }

    fun getCompanyDetail(companyID: String) {
        viewState.showErrorScreen(false)
        viewState.showProgress(true)
        requestCompanyDetail(companyID)
    }

    private fun requestCompanyDetail(companyID: String) {
        val request = restApi.company.getCompanyDetail(companyID)
            .compose(RxUtils.singleTransformers())
            .subscribe({ this.successGetCompanyDetails(it) }, { this.handleError(it) })

        unSubscribeOnDestroy(request)
    }

    private fun successGetCompanyDetails(companyDetail: CompanyDetail) {
        viewState.showProgress(false)
        viewState.showCompanyDetail(companyDetail)
    }

    override fun handleError(throwable: Throwable) {
        super.handleError(throwable)
        viewState.showProgress(false)
        viewState.showErrorScreen(true)
    }

    fun checkAuthUser() {
        auth = FirebaseAuth.getInstance()
        initAuthListener()
        authListener?.let { auth.addAuthStateListener(it) }
    }

    private fun initAuthListener() {
        authListener = FirebaseAuth.AuthStateListener { auth ->
            val user = auth.currentUser
            if (user == null) viewState.showAuth()
            else viewState.addReview()
        }
    }

    fun removeAuthListener() {
        authListener?.let { auth.removeAuthStateListener(it) }
    }

    fun filterReviews(filterType: FilterType, position: String, city: String) {
        this.position = position
        this.city = city
        val sortedReviews = when (filterType) {
            FilterType.RATING -> reviews.sortedByDescending { it.markCompany?.averageMark }
            FilterType.SALARY -> reviews.sortedByDescending { it.markCompany?.salary }
            FilterType.CHIEF -> reviews.sortedByDescending { it.markCompany?.chief }
            FilterType.WORKPLACE -> reviews.sortedByDescending { it.markCompany?.workplace }
            FilterType.CAREER -> reviews.sortedByDescending { it.markCompany?.career }
            FilterType.COLLECTIVE -> reviews.sortedByDescending { it.markCompany?.collective }
            FilterType.BENEFITS -> reviews.sortedByDescending { it.markCompany?.socialPackage }
            FilterType.POPULARITY -> reviews.sortedByDescending { it.like }
        }
        viewState.setFilterIcon(R.drawable.ic_filter_applied)
        if (position.isEmpty() && city.isEmpty()) {
            if (sortedReviews.isNotEmpty()) viewState.showReviews(sortedReviews)
            else viewState.showEmptyReviews()
            return
        }
        var filteredReviews = emptyList<Review>()
        if (position.isNotEmpty() && city.isNotEmpty()) {
            filteredReviews = sortedReviews
                .filter { it.position.equals(position, true) }
                .filter { it.city.equals(city, true) }
        } else if (position.isNotEmpty() && city.isEmpty()) {
            filteredReviews = sortedReviews
                .filter { it.position.equals(position, true) }
        } else if (position.isEmpty() && city.isNotEmpty()) {
            filteredReviews = sortedReviews
                .filter { it.city.equals(city, true) }
        }
        if (filteredReviews.isNotEmpty()) viewState.showReviews(filteredReviews, true)
        else viewState.showEmptyReviews(true)
    }

    fun filterClick() {
        viewState.showFilterDialog(position, city)
    }
}
